### PR TITLE
develop/ipmi-console-robust

### PR DESCRIPTION
**Pull request by WinnieXU1**
1. Catch the exception when node instance has not started BMC or no workspace.
2. Add tests:
   a. start and stop specified ipmi-console;
   b. start ipmi-console when node instance has not started BMC;
   c. start ipmi-console when node instance has no workspace.

### CODE REVIEW

**Comment by emcinfrasim** [comment]
Can one of the admins verify this patch?

File: infrasim/console.py [hunk 1]
```diff
@@ -88,11 +88,11 @@ def run(self):
                 continue
 
 
-def _start_console():
+def _start_console(instance="default"):
```
Comment (turtle-fly) [inline infrasim/console.py]: It comes to me that `instance` can be a environment variable, so you dont' have to pass it to these functions.

File: infrasim/ipmicons/common.py [hunk 2]
```diff
@@ -64,6 +74,12 @@ def init_env(instance):
     have a plan to give instance name to ipmi-console so that it can be attached to
     target vBMC instance.
     """
+    if not Workspace.check_workspace_exists(instance):
```
Comment (turtle-fly) [inline infrasim/ipmicons/common.py]: Good reuse.

File: infrasim/ipmicons/common.py [hunk 3]
```diff
@@ -64,6 +74,12 @@ def init_env(instance):
...
+        raise IpmiError("Warning: there is no node {} workspace. Please start node {} first.".format(instance, instance))
+    output = run_command("infrasim node status")
+    if not (output[0] == 0 and "stop" not in output[1]):
```
Comment (turtle-fly) [inline infrasim/ipmicons/common.py]: Condition here is not rigorous. When vbmc start but qemu stop, this statement here may fail to judge real condition.

File: test/functional/test_ipmi_console.py [hunk 4]
```diff
@@ -92,6 +92,48 @@ def test_start_stop_default_ipmi_console(self):
...
+        assert 'ipmi-console start {}'.format(self.node_name) in output
+        self.assertEqual(returncode1, 0)
+        assert 'ipmi-console start {}'.format(self.node_name) not in output1
+
+    def test_start_ipmi_console_not_start_bmc(self):
+        self.node_name = "default"
+        self.node_workspace = os.path.join(config.infrasim_home, self.node_name)
+        os.system("infrasim node start")
+        os.system("infrasim node stop")
+        os.system("ipmi-console start")
```
Comment (turtle-fly) [inline test/functional/test_ipmi_console.py]: Output of this command should also be captured and the warning should be asserted here.

File: test/functional/test_ipmi_console.py [hunk 5]
```diff
@@ -92,6 +92,48 @@ def test_start_stop_default_ipmi_console(self):
...
+        ipmi_start_cmd = 'ps ax | grep ipmi-console'
+        returncode, output = run_command(ipmi_start_cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        os.system("ipmi-console stop")
+        self.assertEqual(returncode, 0)
+        assert 'ipmi-console start' not in output
+
+    def test_start_ipmi_console_no_workspace(self):
+        self.node_name = "default"
+        self.node_workspace = os.path.join(config.infrasim_home, self.node_name)
+        os.system("ipmi-console start")
```
Comment (turtle-fly) [inline test/functional/test_ipmi_console.py]: Warning in this command's output should be verified as well.

File: test/functional/test_ipmi_console.py [hunk 6]
```diff
@@ -92,6 +92,54 @@ def test_start_stop_default_ipmi_console(self):
...
+        ipmi_start_cmd = 'ps ax | grep ipmi-console'
+        returncode, output = run_command(ipmi_start_cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        os.system("ipmi-console stop")
+        self.assertEqual(returncode, 0)
+        assert 'ipmi-console start' not in output
+
+    def test_start_ipmi_console_no_workspace(self):
+        self.node_name = "default"
+        self.node_workspace = os.path.join(config.infrasim_home, self.node_name)
+        ipmi_start_cmd =("ipmi-console start")
```
Comment (turtle-fly) [inline test/functional/test_ipmi_console.py]: Interesting... ("some string") actually gives a string instead of a tuple

**Comment by turtle-fly** [comment]
retest this please

**Comment by turtle-fly** [comment]
I'm OK :+1: 
@XiaowenJiang @xiar @fub2 please help review

File: test/functional/test_ipmi_console.py (hunk 6 above)
Comment (XiaowenJiang) [inline test/functional/test_ipmi_console.py]: why to add this bracket?

**Comment by XiaowenJiang** [comment]
retest this please

**Comment by WinnieXU1** [comment]
I have removed the bracket @turtle-fly

**Pull request merged**